### PR TITLE
Switch leader scheduler to use PoH ticks instead of Entry height

### PR DESCRIPTION
**Pull request by carllin**
Moved leader scheduler to PoH, precursor to changes/refactoring for the rest of consensus.

### CODE REVIEW

**Comment by rob-solana** [comment]
this should be counting Ticks, not PoH num_hashes

a Tick is defined as an Entry with an empty transactions member.

**Comment by aeyakovenko** [comment]
I think Ticks are the right thing to count, since the network can set the tick “difficulty” in terms of num hashes. But you might want to base this of a Ticks per second global somewhere.

**Comment by rob-solana** [comment]
thinking this through some more...  banking_stage and poh_recorder need to be restructured to emit ticks on specific PoH counts, and to restrict an Entry (i.e. one with transactions) to num_hashes = 1, right?

This code currently incorporates some number of PoH counts within an Entry:
https://github.com/solana-labs/solana/blob/d52e6d01ecaca519340791e14560f1dc714fa2d4/src/poh_recorder.rs#L55

**Comment by aeyakovenko** [comment]
@rob-solana we can fix that later.  or they need to be close enough to the right number of hashes per tick.

File: src/banking_stage.rs [hunk 1]
```diff
@@ -135,22 +168,47 @@ impl BankingStage {
...
             }
-            poh.tick()?;
+            match poh.tick() {
+                Ok(height) if Some(height) == poh.max_tick_height => {
+                    // CASE 1: We were successful in recording the last tick, so exit
+                    return Ok(Some(BankingStageReturnType::LeaderRotation));
+                }
+                Ok(_) => (),
+                Err(Error::PohRecorderError(PohRecorderError::MaxHeightReached)) => {
+                    // CASE 2: Somebody beat us by recording the last tick, so we can just exit
```
[the remark this answers is not in the document]
Comment (carllin) [reply src/banking_stage.rs]: oops that was a remnant from when I was counting num_hashes and both ticks and entries could increment the number of hashes.

File: src/banking_stage.rs [hunk 2]
```diff
@@ -410,4 +505,23 @@ mod tests {
...
+        let max_poh = 10;
+        let (banking_stage, _entry_receiver) = BankingStage::new(
+            &bank,
+            verified_receiver,
+            Default::default(),
+            &bank.last_id(),
+            0,
+            Some(max_poh),
+        );
+        assert_eq!(
```
[the remark this answers is not in the document]
Comment (carllin) [reply src/banking_stage.rs]: More so testing that when the max_tick_height is reached, the banking stage actually exits with the reason BankingStageReturnType::LeaderRotation, which when bubbled up to the tvu + fullnode, will signal to perform a leader rotation. Added a comment.

File: src/entry.rs [hunk 3]
```diff
@@ -48,12 +48,33 @@ pub struct Entry {
...
     /// Creates the next Entry `num_hashes` after `start_hash`.
     pub fn new(start_hash: &Hash, num_hashes: u64, transactions: Vec<Transaction>) -> Self {
-        let num_hashes = num_hashes + if transactions.is_empty() { 0 } else { 1 };
-        let id = next_hash(start_hash, 0, &transactions);
-        let entry = Entry {
-            num_hashes,
-            id,
-            transactions,
+        let entry = {
+            if num_hashes == 0 && transactions.is_empty() {
```
[the remark this answers is not in the document]
Comment (carllin) [reply src/entry.rs]: I think it's been rooted out from everything except the first genesis entry in mint.create() :)

**Pull request merged**